syncstorage/storage/sql/__init__.py

```diff
@@ -593,55 +670,91 @@ def delete_item(self, session, userid, collection, item):
+                                             "total_affected": 0,
+                                             "grace_period": 3 * 60 * 60}}
+
+        total_batches_purged = 0
+        batches_purged = max_per_loop + 1
         is_incomplete = False
-        for table in sorted(tables):
-            logger.info("Purging expired items from %s", table)
-            num_iters = 1
-            num_affected = 0
-            rowcount = session.query("PURGE_SOME_EXPIRED_ITEMS", {
-                "bso": table,
-                "grace": grace_period,
-                "maxitems": max_per_loop,
-            })
-            while rowcount > 0:
-                num_affected += rowcount
-                logger.debug("After %d iterations, %s items purged",
-                             num_iters, num_affected)
-                num_iters += 1
-                if num_iters > 100:
-                    logger.debug("Too many iterations, bailing out.")
-                    is_incomplete = True
-                    break
-                rowcount = session.query("PURGE_SOME_EXPIRED_ITEMS", {
-                    "bso": table,
-                    "grace": grace_period,
-                    "maxitems": max_per_loop,
-                })
-            logger.info("Purged %d expired items from %s",
-                        num_affected, table)
-            total_affected += num_affected
+
+        for table_type in ("bso", "batch_upload_items"):
+            tabula = table_sets[table_type]
+            if table_type == "batch_upload_items":
+                # Tidy up the batch_uploads table first
+                if batches_purged <= max_per_loop:
+                    with self._get_or_create_session() as session:
+                        batches_purged = session.query("PURGE_BATCHES", {
+                            "grace": tabula["grace_period"],
+                            "maxitems": max_per_loop,
+                        })
+                    total_batches_purged += batches_purged
+
+            if not self.dbconnector.shard:
+                tables = set((table_type,))
+            else:
+                shard_func = self.dbconnector.get_bso_table
+                if table_type == "batch_upload_items":
+                    shard_func = self.dbconnector.get_batch_item_table
+                tables = set(shard_func(i).name
+                             for i in xrange(self.dbconnector.shardsize))
+                assert len(tables) == self.dbconnector.shardsize
+            # Purge each table in turn, summing rowcounts.
+            # We set an upper limit on the number of iterations, to avoid
+            # getting stuck indefinitely on a single table.
+            for table in sorted(tables):
+                logger.info("Purging expired items from %s", table)
+                num_iters = 1
+                num_affected = 0
+                with self._get_or_create_session() as session:
```

Comment:
(If it *is* helping performance, it may be worth leaving a comment about why it's important to re-create the session on each iteration.  Future me is very likely to forget that this was important.)